[`Core generation`] Adds support for static KV cache

ArthurZucker opened the pull request:
# ~4x speedups with cuda graphs! 🥳 

Currently getting  ~4x speedups compare to dynamic cache with torch.compile for a single forward pass (agnostic to batch but faster for smaller batch) 

Forward is very very very fast, but materializing the input costs a bit!
~10ms / forward is what we get to! 

- [x] Refactors the way we deal with attention mask:
    - causal and padding are separated 
    - does not rely on the `past_key_values`
    - merged in 2 line. No attention mask utils are needed, no extra complicated logic all explicit
    - LlamaAttention is not self contained, this added 20% overhead in a simple forward
    - Gets rid of the entire `mask_attn_utils` 😄 
- [x] Save the cache class in the generation config
- [x] Init the cache with the batch size (from the generate call) and the `max_length` from the generation config (taking max_new_tokens) into account
- [x] torch.compile

## Benchmark using [af097af](https://github.com/huggingface/transformers/pull/27931/commits/af097af7582de01fcf73856e7fa37be972e96456)


## Use it in generate:
Use this: EDIT: TO COME


## Failing test left for @gante
Related to the fact that I don't return `past_key_values` / is None so the `test_new_cache_format` fails. I don't want to dive in this. 


fixes #28075 , fixes #28610, fixes #28190

ArthurZucker committed:
initial commit

ArthurZucker committed:
lol

ArthurZucker committed:
nits

ArthurZucker committed:
nits nits nits nits nits

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27931). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,91 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
```
You'll have to pass the attention_mask here as well I think

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,91 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        if len(self.key_cache) + 1 == self.max_sequence_length:
+            # let's overwrite and roll the cache to support going beyond?
+            self.key_cache[layer_idx][0] = key_states
+            self.key_cache[layer_idx] = torch.roll(self.key_cache[layer_idx],-1,0)
```
Hmm I'd not allow this. `torch.roll` doesn't work well with torch.compile as it dynamically allocates new memory. I'd throw an error instead if the key_cache length is larger then max_length

ArthurZucker replied:
Agreed!

oobabooga commented:
If I understand correctly, this PR should close the existing gap between inference with transformers + AutoGPTQ and inference with ExLlama, as the VRAM usage would become much more controlled. I'm rooting for it :)

ArthurZucker commented:
Thanks! 🤗

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
some nits and some testing

ArthurZucker committed:
nits

ArthurZucker committed:
Wrong implementation but creates good masks in general and is pretty simple

ArthurZucker committed:
what seems to work for now

ArthurZucker committed:
nites

ArthurZucker committed:
re-init cache

ArthurZucker committed:
make it automatic

ArthurZucker committed:
nits and nits

ArthurZucker committed:
more nits

ArthurZucker committed:
nits

ArthurZucker committed:
nits

ArthurZucker committed:
more nits

patrickvonplaten commented:
Exciting PR!

ArthurZucker committed:
nits

ArthurZucker committed:
fastest working cache for now

ArthurZucker committed:
also include the attention mask

gante reviewed (commented):
2-3x speedups look super exciting 🔥🔥 

High-level questions/comments:
1. `StaticCache` inherits from `nn.Module`. Is there a particular reason why?
2. This PR adds a new generation kwarg, to select the cache type. What we already have in place is the ability to pass an initialized cache instance to `generate`, through `past_key_values=cache_instance`. Is this not possible? (fewer generate kwargs = 💛)
3. Coupling the cache and the attention mask is (IMO) a bad idea, as it a) defeats the point of isolating the cache into separate plug-and-play classes with a standardized API and b) would require more modelling changes. In TF/JAX we delegate the attention mask handling job to generate itself (e.g. [here](https://github.com/huggingface/transformers/blob/ee2482b6f8eca320f1eff112673adae215dbc927/src/transformers/generation/tf_utils.py#L1294C9-L1294C29)), and the model simply consumes the inputs as usual.

Possibly related to some points above, and to make sure we're on the same page. Our main target should be to compile `model.forward`, and not necessarily `model.generate`: the former has the bulk of the computation, and it would still enable streaming.

ArthurZucker commented:
1. This was some debug / tests but removed hehe
2. The idea is to be able to save this and track usage + just make the api seamless. You can pass the cache instance but it should be able to be deduced. It's not a generation kwargs for me but a generation_config argument rather than that 
3. Attention mask is mostly what is left to adress TBH, because the control flow introduced are non trivial that is why I have a seperate PR that tried to take inspiration for JAX. Still working on it!

ArthurZucker committed:
updates

ArthurZucker committed:
current state

ArthurZucker committed:
working code

ArthurZucker committed:
dummy mask for now

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
Merge branch 'static-cache' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

hhllxx1121 commented:
How do I use this PR on npu device, here is my demo with errors.
```python
import torch
import torch_npu
from transformers import LlamaForCausalLM, LlamaTokenizer, TextStreamer



tokenizer = LlamaTokenizer.from_pretrained(
    "../Llama-2-7b-chat-hf",
    device_map="npu:2"
)
res = tokenizer("hello")

llama_model = LlamaForCausalLM.from_pretrained(
    "../Llama-2-7b-chat-hf",
    device_map="npu:2"
)
streamer = TextStreamer(tokenizer)


print("load successful")
while True:
    ins = input("user: ")
    res = tokenizer.encode(ins, return_tensors="pt").to("npu:2")
    print(res.dtype)
    outputs = llama_model.generate(
        inputs=res,
        streamer=streamer,
        max_new_tokens=10,
    )
```

Traceback (most recent call last):
  File "test.py", line 27, in <module>
    streamer=streamer,
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/generation/utils.py", line 1790, in generate
    return self.sample(
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/generation/utils.py", line 2887, in sample
    outputs = self(
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 1206, in forward
    outputs = self.model(
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 1093, in forward
    layer_outputs = decoder_layer(
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 820, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/conda/envs/PyTorch2.1.0/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 438, in forward
    raise ValueError(
ValueError: Attention mask should be of size (1, 1, 2, 2), but is torch.Size([1, 2])

ArthurZucker commented:
This PR is still very much in draft mode. A script can be found in the tests, but does not support everything yet! Would recommend you to wait a tad bit, should be in a better state by the end of the week

ArthurZucker committed:
a better design

ArthurZucker committed:
some fix

ArthurZucker committed:
make outputs match

ArthurZucker committed:
fastest yet

ArthurZucker committed:
remove chunck qkv

ArthurZucker committed:
cleanup

ArthurZucker committed:
some test

ArthurZucker committed:
goat changes

ArthurZucker committed:
nits

ArthurZucker committed:
dynamic was not working anymore

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1009,12 +1035,8 @@ def forward(
                 )
                 use_cache = False
 
-        past_key_values_length = 0
-        if use_cache:
-            use_legacy_cache = not isinstance(past_key_values, Cache)
-            if use_legacy_cache:
-                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
-            past_key_values_length = past_key_values.get_usable_length(seq_length)
```
let' s get rid of this dependency on the cache class / object.

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1029,20 +1051,6 @@ def forward(
-        else:
-            # 4d mask is passed through the layers
-            attention_mask = _prepare_4d_causal_attention_mask(
-                attention_mask, (batch_size, seq_length), inputs_embeds, past_key_values_length
```
let' s get rid of this as well

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -403,29 +404,39 @@ def forward(
+            causal_mask = self.causal_mask[None, cache_positions, :past_key_value.get_max_length()].repeat(bsz, 1, 1)
+            # mask out padding and unsqueeze in head position
+            causal_mask[:,:q_len,:kv_seq_len].mul_(attention_mask[:,None,:])
+            causal_mask = causal_mask.unsqueeze(1)
```
3 lines are enough to go from 2d to 4d mask....

ArthurZucker committed:
cache reverts

ArthurZucker committed:
small nits

ArthurZucker committed:
sdpa

ArthurZucker committed:
Merge branch 'static-cache' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
make sure sdpa passed

ArthurZucker committed:
nit

ArthurZucker committed:
cleqnups

younesbelkada commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,77 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+
+class StaticCache(Cache):
+
+    def __init__(self, config: PretrainedConfig, max_batch_size, device = "cuda:2") -> None:
```
```suggestion
    def __init__(self, config: PretrainedConfig, max_batch_size, device = "cpu") -> None:
```
maybe better to init on CPU?

ArthurZucker replied:
That's for debugging for now!

younesbelkada commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,77 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+
+class StaticCache(Cache):
+
+    def __init__(self, config: PretrainedConfig, max_batch_size, device = "cuda:2") -> None:
```
```suggestion
    def __init__(self, config: PretrainedConfig, max_batch_size, device) -> None:
```

younesbelkada commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,77 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        self.max_sequence_length = config.max_position_embeddings if config.max_sequence_length is None else config.max_sequence_length 
+        self.head_dim = config.hidden_size // config.num_attention_heads
+        self.num_heads = config.num_attention_heads
+        self.dtype = config.torch_dtype if config.torch_dtype  is not None else torch.float16
```
```suggestion
        self.dtype = config.torch_dtype if config.torch_dtype  is not None else torch.float32
```

younesbelkada replied:
to make it more universal I would init here everything in fp32

younesbelkada commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +322,77 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+            A tuple containing the updated key and value states.
+        """
+        position_ids = cache_kwargs.get("position_ids")
+        # position_ids = cache_kwargs.get("position_ids")[0] is faster?
+        # position_ids = torch.arange(self.seen_tokens, self.seen_tokens + key_states.shape[-2], device=key_states.device)
```
```suggestion
```

younesbelkada reviewed (commented):
I left some early comments , and thanks for all the offline discussion! Left some comments as requested offline

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -1322,6 +1322,14 @@ def generate(
+        }
+        if generation_config.cache_implementation in ALL_CACHE_CLASSES and not model_kwargs.get("past_key_values", False):
+            cache_cls = ALL_CACHE_CLASSES[generation_config.cache_implementation]
+            self._setup_cache(cache_cls, batch_size)
```
I guess this is the big design change which instantiates a `cache_cls` to each attn class. I understand why this is necessary and think it's ok to do this for the static cache, but I think we should not do it for the dynamic or other caches because it'll prevent the model from being used in a multi-threading environment. We can simply say that when enabling the static cache, multi-threading is not possible, but we should probably keep the dynamic cache the way it is so that the model can be used in a multi-threading environment

ydshieh replied:
IIRC, only if `generation_config.cache_implementation == "static"` will have `model_kwargs["past_key_values"] = cache_cls(self.config, max_batch_size=batch_size)`. So I'm not seeing the problem, unless @ArthurZucker  plan to extend `ALL_CACHE_CLASSES` in the future.

BTW, what's the issue with the  multi-threading when a `cache_cls` is used?

ydshieh commented on src/transformers/generation/utils.py:
```diff
@@ -1322,6 +1322,14 @@ def generate(
         )
         batch_size = inputs_tensor.shape[0]
 
+        ALL_CACHE_CLASSES = {
+            "static": StaticCache
+        }
+        if generation_config.cache_implementation in ALL_CACHE_CLASSES and not model_kwargs.get("past_key_values", False):
+            cache_cls = ALL_CACHE_CLASSES[generation_config.cache_implementation]
```
I am not seeing any usage of `ALL_CACHE_CLASSES` except here. And the block here could be simply like

```python
if generation_config.cache_implementation == static and not ....:
    cache_cls = StaticCache
    model_kwargs["past_key_values"] = cache_cls(self.config, max_batch_size=batch_size)
```

Not sure if you have future usage of `ALL_CACHE_CLASSES` in mind so you make the decision here though.

ydshieh commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -865,6 +838,9 @@ def _init_weights(self, module):
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
 
+    def _setup_cache(self, cache_cls, max_batch_size):
+        for layer in self.model.layers:
+            layer.self_attn.past_key_value = cache_cls(self.config, max_batch_size, device=layer.self_attn.o_proj.weight.device)
```
well, if we ever will have other new cache classes in the future, it's fine here and also having `ALL_CACHE_CLASSES` in `utils.py`.

ydshieh commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -315,8 +280,12 @@ def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
         self._init_rope()
+        
+        # register a causal mask to separate causal and padding mask creation. Merging happends in the attention class
+        causal_mask = torch.tril(torch.full((self.max_position_embeddings, self.max_position_embeddings), fill_value=1))
+        self.register_buffer("causal_mask", causal_mask, persistent=False)
```
It looks to me that this `causal_mask` would be the same (shape and its values) across all layers. Maybe we can create a **single buffer** in `LlamaModel`, and pass it down to `LlamaAttention`  at `__init__` time and then set `self.causal_mask = causal_mask`.

ArthurZucker replied:
Yes and no, this would not make the attention class self contained, and could have issue with parallel device as the memory will be shared when running attnetion. I'll keep it as is as it's a fairly small tensor.

gante reviewed (commented):
The current version of the static cache has the following API changes:
1. The cache class holds the cache for one layer, as opposed to for all layers;
2. The cache is now being held by the model as an attribute (`layer.self_attn.past_key_value`), as opposed to being passed around. This means that initialization must also be a model method.

Are these hard requirements to extract the large speedups? 
👉 If yes, then we should open a follow-up PR after this one to standardize the use of caches according to this structure, while preserving full retrocompatibility. We need to be aware that it will take some effort to propagate all this across the codebase 💪 
👉 If not, then I'd suggest to keep the existing structure. Not because it is better/worse, but because it will greatly reduce the scope of needed changes

Other than that, aside from a few nits that I'll review when we're all happy with the overall structure, it has my green light

ArthurZucker committed:
cleanup

ArthurZucker committed:
nits

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
pass sdpa

ArthurZucker committed:
make sure dynamic is BC

ArthurZucker committed:
update check on the attn weight

ArthurZucker committed:
Merge branch 'static-cache' of https://github.com/huggingface/transformers into static-cache

ArthurZucker committed:
faster?

ArthurZucker committed:
add `_reset_cache`

ArthurZucker committed:
Merge branch 'static-cache' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
nit

ArthurZucker committed:
Merge branch 'static-cache' of https://github.com/huggingface/transformers into static-cache

ArthurZucker committed:
nit

ArthurZucker committed:
Merge branch 'static-cache' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
merges

ArthurZucker committed:
Styling

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1009,41 +983,14 @@ def forward(
-        elif self._use_sdpa and not output_attentions:
-            # output_attentions=True can not be supported when using SDPA, and we fall back on
-            # the manual implementation that requires a 4D causal mask in all cases.
-            attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
-                attention_mask,
-                (batch_size, seq_length),
-                inputs_embeds,
-                past_key_values_length,
-            )
```
cc @fxmarty I think I'd rather we disable `mem_efficient` than have to fix this only for float32, as float32 is not really used that often.

fxmarty replied:
No. This issue exists as well in bf16.

ArthurZucker committed:
nites

ArthurZucker committed:
revert some BC breaking changes

ArthurZucker committed:
make all tests pass

ArthurZucker committed:
torch long not float for attention mask

ArthurZucker committed:
try to remove the guard

ArthurZucker committed:
BC

ArthurZucker committed:
even more cleanup

ArthurZucker committed:
fix `past_key_value.get_usable_length(kv_seq_len, self.layer_idx)`

ArthurZucker committed:
pushh a fast version

ArthurZucker committed:
what actually works

ArthurZucker committed:
no contigious()

ArthurZucker committed:
push for eager as well

ArthurZucker committed:
simplest and best way to do it yet

ArthurZucker committed:
merge

ArthurZucker committed:
style

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into static-cache

ArthurZucker committed:
dix dtype

ArthurZucker committed:
fix dtype issues

ArthurZucker committed:
nits

fxmarty commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1026,22 +985,23 @@ def forward(
+                
+        causal_mask = self.causal_mask[None, None, :, :].repeat(batch_size, 1, 1, 1)
+        if attention_mask is not None and attention_mask.dim() == 2:
+            paddin_mask = causal_mask[..., : attention_mask.shape[-1]].eq(False) * attention_mask[:, None, None, :].eq(
+                False
+            )
+            causal_mask[..., : attention_mask.shape[-1]] = causal_mask[..., : attention_mask.shape[-1]].masked_fill(
+                paddin_mask, torch.finfo(inputs_embeds.dtype).min
```
In case SDPA is used, you need to make sure that there is no row fully masked: https://github.com/pytorch/pytorch/issues/110213 (which is what `_prepare_4d_causal_attention_mask_for_sdpa` does)

ArthurZucker replied:
I don't want to :imp: pretty sure it's only for float32?

fxmarty replied:
bf16 and fp32 have the same number of bits for the range.

ArthurZucker replied:
I realllllly don't want to :(

ArthurZucker replied:
I'll see what I can do to make it as simple as possible

ArthurZucker replied:
Turns out `causal_mask.mul_(~torch.eq(causal_mask, causal_mask.min()).all(dim=-1)[..., None])` is enough

ArthurZucker commented:
cc @oobabooga the PR is more or less ready, would love to get your feedback from using it in `oobabooga`! (this needs a bit of custom work to compile only the decoding step, but should already help!)

ArthurZucker committed:
nit

fxmarty commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1026,22 +985,23 @@ def forward(
+                
+        causal_mask = self.causal_mask[None, None, :, :].repeat(batch_size, 1, 1, 1)
+        if attention_mask is not None and attention_mask.dim() == 2:
+            paddin_mask = causal_mask[..., : attention_mask.shape[-1]].eq(False) * attention_mask[:, None, None, :].eq(
+                False
+            )
+            causal_mask[..., : attention_mask.shape[-1]] = causal_mask[..., : attention_mask.shape[-1]].masked_fill(
+                paddin_mask, torch.finfo(inputs_embeds.dtype).min
```
```suggestion
            paddin_mask = causal_mask[..., : attention_mask.shape[-1]].ne(0.) * attention_mask[:, None, None, :].ne(0.)
            causal_mask[..., : attention_mask.shape[-1]] = causal_mask[..., : attention_mask.shape[-1]].masked_fill(
                paddin_mask, torch.finfo(inputs_embeds.dtype).min
```

and rename `paddin_mask` to `padding_mask`

ArthurZucker committed:
support export to torchscript

ArthurZucker committed:
Credit helpers

Co-authored-by: fxmarty <9808326+fxmarty@users.noreply.github.com>
Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

ArthurZucker committed:
nits

fxmarty commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1026,22 +985,21 @@ def forward(
-        elif self._use_sdpa and not output_attentions:
-            # output_attentions=True can not be supported when using SDPA, and we fall back on
-            # the manual implementation that requires a 4D causal mask in all cases.
-            attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
```
By the way, an other reason why it is not good to drop it like that is that in the implementation you have you will never dispatch to flash attention through SDPA since you always pass a custom attention mask, can't we drop the mask for the [relevant cases](https://github.com/huggingface/transformers/blob/ec29d25d9f7109f3fdaadfa51515eb6745a136ba/src/transformers/modeling_attn_mask_utils.py#L371-L381)?

ArthurZucker replied:
Agreed that we'll keep dropping the mask. Think it suffice to check that the first mask of all batch is always 0 -> set to None

ArthurZucker committed:
handle SDPA edge cases

ArthurZucker committed:
handle sdpa quircks

ArthurZucker committed:
revert performance break

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1399,6 +1403,19 @@ def generate(
             generation_config.max_length = generation_config.max_new_tokens + input_ids_length
+
+        # if we don't pass `past_key_values` and a cache_implementation is specified
+        if generation_config.cache_implementation in ALL_COMPILE_CACHE_CLASSES_MAPPING and not model_kwargs.get(
```
```suggestion
        if generation_config.cache_implementation in NEED_SETUP_CACHE_CLASSES_MAPPING and not model_kwargs.get(
```
nit: let's use a more flexible naming :) (compilable caches need setup, but the reverse may not be true)

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -392,38 +350,25 @@ def forward(
-                    "with a layer index."
-                )
-            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)  # add what was seen
```
```suggestion
            past_seen_tokens = past_key_value.get_usable_length(kv_seq_len, self.layer_idx)  # add what was seen
            kv_seq_len += past_seen_tokens
        else:
            past_seen_tokens = 0
```
nit: I think it's more readable this way :) (deleting the `past_seen_tokens = kv_seq_len - key_states.shape[-2]` line below)

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -392,38 +350,25 @@ def forward(
+        new_cache_positions = torch.arange(past_seen_tokens, past_seen_tokens + q_len, device=key_states.device)
         if past_key_value is not None:
-            cache_kwargs = {"sin": sin, "cos": cos}  # Specific to RoPE models
+            cache_kwargs = {"sin": sin, "cos": cos, "position_ids": new_cache_positions}  # Specific to RoPE models
```
```suggestion
            # sin and cos are specific to RoPE models; position_ids needed for the static cache
            cache_kwargs = {"sin": sin, "cos": cos, "position_ids": new_cache_positions}
```
nit: more precise comment

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -392,38 +350,25 @@ def forward(
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
+        past_seen_tokens = kv_seq_len - key_states.shape[-2]
```
```suggestion
```

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -699,28 +636,34 @@ def forward(
 
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
+        past_seen_tokens = kv_seq_len - key_states.shape[-2]
```
```suggestion
```

gante reviewed (approved):
Many tiny nits, but should be quick to solve! Let's get this merged 🔥

oobabooga commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1026,22 +994,25 @@ def forward(
+
+        causal_mask = self.causal_mask[None, None, :, :].repeat(batch_size, 1, 1, 1)
+        if attention_mask is not None and attention_mask.dim() == 2:
+            padding_mask = causal_mask[..., : attention_mask.shape[-1]].eq(0.0) * attention_mask[:, None, None, :].eq(
```
I haven't been able to test the PR because this line fails with error

```
    padding_mask = causal_mask[..., : attention_mask.shape[-1]].eq(0.0) * attention_mask[:, None, None, :].eq(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

I'm calling `model.generate` for llama-2-7b-hf. I Initialize the static cache with

```
    model.generation_config.cache_implementation = "static"
    model.generation_config.max_length = 4096
```

after loading it. Am I doing something wrong?

ArthurZucker replied:
Nope! I’ll check the device thing again, probably needs a nit! Could you share the repro? (Multi GPU setting maybe?)

ArthurZucker replied:
Fixing today

oobabooga replied:
Here is a simple reproduction

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# Load pre-trained model and tokenizer
model = AutoModelForCausalLM.from_pretrained(
    'models/llama-2-7b-hf',
    torch_dtype=torch.float16
).cuda()

# Set generation configuration parameters
model.generation_config.cache_implementation = "static"
model.generation_config.max_length = 4096

# Load tokenizer
tokenizer = AutoTokenizer.from_pretrained('models/llama-2-7b-hf')

# Encode input text
input_ids = tokenizer.encode('This is a test.', return_tensors='pt')

# Generate text based on input
output = model.generate(
    input_ids,
    temperature=0.7,
    max_new_tokens=32
)

# Print the generated output
print(output)
```

The error also happens if those lines are commented-out, so it doesn't seem to be an issue with static cache specifically

```python
model.generation_config.cache_implementation = "static"
model.generation_config.max_length = 4096
```

ArthurZucker replied:
Yep it's the new way of handling causal mask.
Could you use `device_map = "cuda"` instead of `.cuda()`?

ArthurZucker replied:
Also generating this way will not use the compiled version, just the static cache. A follow up PR will support it natively. For now something like 
```
....
compiled_forward = torch.compile(model)
def model_forward(input_ids, **kwargs):
    if input_ids.shape[1] == 1:
        return compiled_forward(input_ids, **kwargs)
    return model(input_ids, **kwargs)
model.__call__ = model_forward
```

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +324,74 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        self.dtype = config.torch_dtype if config.torch_dtype is not None else torch.float32
+
+        cache_shape = (max_batch_size, self.num_heads, self.max_cache_len, self.head_dim)
+        self.key_cache: torch.Tensor = torch.zeros(cache_shape, dtype=self.dtype, device=device)
+        self.value_cache: torch.Tensor = torch.zeros(cache_shape, dtype=self.dtype, device=device)
```
Leave it to you for a follow up PR 😉

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -865,6 +807,25 @@ def _init_weights(self, module):
+
+        for layer in self.model.layers:
+            layer.self_attn.past_key_value = cache_cls(
+                self.config, max_batch_size, max_cache_len, device=layer.self_attn.o_proj.weight.device
```
good catch

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1026,22 +994,25 @@ def forward(
-            attention_mask = _prepare_4d_causal_attention_mask(
-                attention_mask, (batch_size, seq_length), inputs_embeds, past_key_values_length
+            self.register_buffer("causal_mask", torch.triu(causal_mask, diagonal=1), persistent=False)
+            logger.warning(
+                "You are going above the `max_position_embedding` you should set `max_position_embedding` accordingly. This will no longer be supported"
+                " in transformers v4.40"
+            )
```
Because of DynamicRopeScaling? Might be good to have no warnings at all as well TBH

gante replied:
agreed

ArthurZucker commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -656,28 +656,34 @@ def forward(
+        past_key_value = getattr(self, "past_key_value", past_key_value)
         if past_key_value is not None:
-            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)  # add what was seen
```
yep! Copied from will deal with that

ArthurZucker committed:
Apply suggestions from code review

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ArthurZucker committed:
fix merges

ArthurZucker committed:
revert removing ```
    def to_legacy_cache(self):
        """Dummy function for BC should not be used"""
        return None
```
as it is required

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +324,74 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+
+    def to_legacy_cache(self):
+        """Dummy function for BC should not be used"""
+        return None
```
We actually need it because of the strange `use_legacy_cache` logic and conversion.

ArthurZucker committed:
add another test

ArthurZucker committed:
update test

ArthurZucker committed:
Merge branch 'static-cache' of https://github.com/huggingface/transformers into static-cache

ArthurZucker committed:
use a model that is not protected

ArthurZucker committed:
only test generation

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -320,3 +324,74 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+        self.max_cache_len = config.max_position_embeddings if max_cache_len is None else max_cache_len
+        self.head_dim = config.hidden_size // config.num_attention_heads
+        self.num_heads = config.num_attention_heads
+        self.dtype = config.torch_dtype if config.torch_dtype is not None else torch.float32
```
If the cache is a nn.Module it should be moved as well no? I'll try that

ArthurZucker committed:
update the cache utils to define the position_ids in the cache class

ArthurZucker committed:
fix static cache

ArthurZucker committed:
add subtest to llama tests

ArthurZucker committed:
update testing suite

ArthurZucker committed:
nuke whatever we can

ArthurZucker committed:
smthing wrong with cache

ArthurZucker committed:
nit

younesbelkada commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -699,28 +640,32 @@ def forward(
+            kv_seq_len += past_seen_tokens
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
 
+        new_cache_positions = torch.arange(past_seen_tokens, past_seen_tokens + q_len, device=key_states.device)
```
I did not benchmarked that yet, but it seems this might create a considerable overhead for FA2 models as for each attention module on each layer you need to re-create `new_cache_positions` and this might become quite an overhead for large sequence length. I will do some benchmarks and get back

ArthurZucker replied:
Yes, but hear me out. It's a fairly small tensor. Either you create it to slice, or you don't but you slice with a list, which will induce device copy. 
So we have 1 solution:
- pass the `cache positions` as inputs. 
- pass the `postiions_ids` as inputs. 
This means we have to keep track of the real positions, and the cache positions. 
Also my benchmark give no overhead for this

younesbelkada commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -656,28 +656,34 @@ def forward(
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
+        past_key_value = getattr(self, "past_key_value", past_key_value)
```
Why the FA2 modules are not updated here whereas the FA2 module is being updated on Llama?

ArthurZucker replied:
Forgot about it! Thanks for catching

younesbelkada commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -656,28 +656,34 @@ def forward(
-                    f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
-                )
+        if (
+            attention_mask is not None and not torch.all(attention_mask[..., 0] == 1) and q_len != 1
```
for large sequence length > 4096, the condition `torch.all(attention_mask[..., 0] == 1)` might create an overhead as well. When I started integrating FA2 with @fxmarty we used a similar condition to check if there was any PAD token, and the overhead was so high that there was no point adding FA2 for inputs with pad tokens. What is the overhead introduced here for large sequence length (just comparing SDPA your branch vs main)

ArthurZucker replied:
No overhead as the same is done in the `_perpare_attention_for_sdpa`

younesbelkada commented on src/transformers/models/mixtral/modeling_mixtral.py:
```diff
@@ -736,28 +736,34 @@ def forward(
+        if (
+            attention_mask is not None and not torch.all(attention_mask[..., 0] == 1) and q_len != 1
+        ):  # user defined causal mask
+            causal_mask = attention_mask[:, :, past_seen_tokens : past_seen_tokens + q_len, : key_states.shape[-2]]
```
My benchmarking script fails here with the following traceback:

```bash
line 1040, in _update_causal_mask
    padding_mask = causal_mask[...,:mask_length].eq(0.0) * attention_mask[:, None, None, :].eq(0.0)
RuntimeError: The size of tensor a (4096) must match the size of tensor b (4097) at non-singleton dimension 3
```
This is because the casual mask is by default created with the same size as `max_positon_embeddings`. I think this assumption can be risky because many users run generation with large context (which is what I do in my benchmarking script), so the PR introduces a breaking behaviour for these users by default. IMO this should be revisited by dynamically re-allocating the `causal_mask` if we go beyond `max_positon_embeddings`. 
Currently running a simple generation with Llama2 with a context that is greater than 4096 fails with this PR, which is possible currently on main

ArthurZucker replied:
No again, it's not breaking if we just take care of it. 
There is a clause for this in `_update_causal_mask` which just needs to allocate a bigger mask. 
Again the overhead is None compared to main for this as we always allocate it. Quite the contrary this means that we dynamically, by stride of `max_embedding_dim` allocate the causal mask needed, instead of always re-creating it in the forward

younesbelkada reviewed (commented):
Thanks a lot for all your work on this @ArthurZucker , I tried to battle test your PR and I think many crucial points needs to be solved before merging.  E.g. doing a simple generation large context fails with this branch, whereas it is possible to do it on main (this is IMO breaking), I left one major question for FA2 that needs to be addressed, currently I can't run the benchmarking script on long contexts because of this issue. 
I believe that we introduced many operations that adds considerable overhead in the case of large sequence length (mainly the cached_position_ids that are computed on each layer, each attention module) - Please have a look at them before merging !

ArthurZucker commented:
I had a look at them, the main pain point here is: 
- the overhead on a generation is ~2 token /s 
- the overhead of not using the `position_ids` tensor in compiled mode is more than 40 tokens / s
s

ArthurZucker committed:
latest changes

ArthurZucker committed:
Merge branch 'main' of https://github.com/huggingface/transformers into static-cache

ArthurZucker committed:
don't use einsum

ArthurZucker committed:
nit

ArthurZucker committed:
remove one unused var

ArthurZucker committed:
update test value

ArthurZucker committed:
let style be happy

ArthurZucker committed:
make sure cache tests are slow

ArthurZucker committed:
slow was removed add it back to test cach utils

ArthurZucker committed:
fix flash_attention_2

ArthurZucker committed:
very small nit

ArthurZucker committed:
revert test change

ArthurZucker committed:
make mistral the default copied from

ArthurZucker committed:
fix copies

ArthurZucker committed:
nits

ArthurZucker committed:
finishup

ArthurZucker committed:
fixup

ArthurZucker committed:
Merge branch 'main' of https://github.com/huggingface/transformers into static-cache

ArthurZucker committed:
skip tests

ArthurZucker commented:
Alternative for ROPE:
```python
class LlamaRotaryEmbedding(nn.Module):
    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
        super().__init__()

        self.dim = dim
        self.max_position_embeddings = max_position_embeddings
        self.base = base
        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2, dtype=torch.int64).float().to(device) / self.dim))
        self.register_buffer("inv_freq", inv_freq, persistent=False)
        # Build here to make `torch.jit.trace` work.
        self._set_cos_sin_cache(
            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
        )

        
    def _set_cos_sin_cache(self, seq_len, device, dtype):
        self.max_seq_len_cached = seq_len
        t = torch.arange(self.max_seq_len_cached, device=device, dtype=torch.int64).type_as(self.inv_freq)

        freqs = torch.outer(t, self.inv_freq)
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.cos_cached = nn.Embedding(self.max_seq_len_cached, 1)
        self.cos_cached.weight.data = emb.cos().to(dtype)
        self.sin_cached = nn.Embedding(self.max_seq_len_cached, 1)
        self.cos_cached.weight.data = emb.sin().to(dtype)

        
    def forward(self, x, position_ids, seq_len=None):
        # x: [bs, num_attention_heads, seq_len, head_size]
        
        return self.cos_cached(position_ids)[0,:,:].to(dtype=x.dtype), self.sin_cached(position_ids)[0,:,:].to(dtype=x.dtype)
```

tsengalb99 commented:
@ArthurZucker when will the next minor release (4.37.3?) of transformers come out? Will this PR be included in 4.37.3?

amyeroberts commented:
@tsengalb99 v4.37.3 would be a micro release. There are reserved for patch fixes which resolve newly introduced breaking changes in the repo or fixing regressions from the last release.

The next minor release is v4.38. We release on a roughly monthly schedule, so would be in around two weeks. 

If you want to use this feature immediately, you can [install from source](https://huggingface.co/docs/transformers/installation#install-from-source).

patrickvonplaten commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -656,28 +659,34 @@ def forward(
-                    f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
-                )
+        if (
+            attention_mask is not None and not torch.all(attention_mask[..., 0] == 1) and q_len != 1
```
(nit)
```py
attention_mask is not None and q_len != 1 and not torch.all(attention_mask[..., 0] == 1)
```

Would be a tad better as it would not run the expensive torch.all computation if `q_len == 1`

ZachariahMarrero commented:
Is this working in general or only for some specific hardware, etc? 
I tried using fresh installs with WSL2, ubuntu, Cuda 11.8 (and again with 12.1) on an RTX A5000

 I installed from source today and tried the test script (https://gist.github.com/ArthurZucker/2dd607c4333ac4c489af30f54a1d8a2d)

 but the model.generate for compiled forward throws runtimeError: attn_bias is not correctly aligned (strideH). attn_bias.stride(1) = 21, and should be a multiple of 8.

tsengalb99 commented:
What is the recommended way to use cuda graphs with this PR? The torch cuda graph wrapper does not appear to be working iirc due to this line `attention_mask is not None and q_len != 1 and not torch.all(attention_mask[..., 0] == 1)`

ArthurZucker commented:
This line is probably gonne be removed in #28937 
I'update once this is ready.
I use torch nightly so torch 2.3!

chauhang commented:
@ArthurZucker Great to see the PR has been merged. I tried with the latest nightlies on H100, and running into errors (full [logs](https://gist.github.com/chauhang/a0dabe50c9958af5c92a3b1a23d1e888))

with sdpa attention: 
`RuntimeError: Event device type CUDA does not match blocking stream's device type CPU.`

with flash_attention_2:
`flash_attn_2_cuda.cpython-310-x86_64-linux-gnu.so: undefined symbol: _ZN3c104cuda9SetDeviceEi`

ArthurZucker commented:
Nice, could you share a full stack trace? 🤗

fxmarty commented:
@chauhang 

> with flash_attention_2 (fails in import itself):

This means that the your local torch version does not match the torch version flash-attn was built against. You would need to uninstall & rebuild flash-attn.

xkszltl commented:
Could you help clarify the removal of comment?
![image](https://github.com/huggingface/transformers/assets/5203025/f5579c0a-362f-4e6c-a28a-52f4c3128293)
It's probably removed by mistake as attn_weights is still not supported in FA2.
We should also add a warning before setting it to False, just like the SDPA counterpart.

paulcx commented:
Hi @ArthurZucker It seems that the increase in VRAM could potentially lead to out of memory (OOM) [comment1](https://github.com/huggingface/transformers/pull/29285#issuecomment-1979879993) [comment2](https://github.com/huggingface/transformers/pull/29285#issuecomment-1980165355), as pointed out in this PR by @danielhanchen 

> It seems like a change was made in another PR which allocates a causal mask of size (16384, 16384) https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L940
> 
> The triu causes the causal mask to upcast to float32, using 16384^2 * 4bytes = 1GB of extra VRAM. We have n^2 * 4 / 1024 / 1024 = 37.25GB in your screenshot, so I'm assuming you're also doing RoPE Scaling to 100K context length? So ie a (100K, 100K) matrix was trying to be created.

Could you please take a look into it?

fxmarty commented:
See https://github.com/huggingface/transformers/pull/29241 which alleviates but does not fix the issue @paulcx

aliencaocao commented:
Does this work for llava? from my testing it doesnt

paulcx commented:
> See https://github.com/huggingface/transformers/pull/29241 which alleviates but does not fix the issue @paulcx 

does this temporary fix work for 200K?

ArthurZucker commented:
No, you have to update the `max_position_embedding`. It is allocating 200k because you set it to 200K while you machine does not support 200K input

ArthurZucker commented:
`max_position_embedding` is only use for the `causal_mask` now, while previously it was used for `sin` and `cos`. In both cases it was to `cache` the maximum number of positions that will be passed to the model. If you have 200K context length, that does not mean you can do inference / training with it!

We can also just remove it, but then you need to allocate the causal_mask at each foward pass

paulcx commented:
@ArthurZucker If I understand correctly, my thought is to lower the max_position_embedding, such as 4096, because it only affects the initial position embedding and caching for 200K. But during inference, lengths approaching 200K will still be calculated, just slower. This workaround, it can ensure normal training and inference for non-200K cases. Is my understanding correct?

ArthurZucker commented:
Yes, the idea is to cache the causal mask for the length supported by your device / corresponding to your training. 
If that does not make sense for you we can think about something else, I'm trying to find the optimal solution here, which for me is to properly rely on the `max_position_embedding` which should only affect the caching of the causal mask! 🤗

ArthurZucker commented:
This was fixed by #29753 ! Sorry @paulcx for the inconvenience. For static cache you should still reduce the max position embedding or it will OOM 😉

paulcx commented:
> static / compile cache

Thank you and great work on new release @ArthurZucker.

Would you mind clarifying the use case of "static / compile cache" in release note? I'm not sure if I understand correctly.

ArthurZucker commented:
It is mostly this: https://gist.github.com/ArthurZucker/af34221def212259b43d55a2811d2dbb, you can get x4 generation speed in transformers with torch compile and static cache!

aliencaocao commented:
> It is mostly this: https://gist.github.com/ArthurZucker/af34221def212259b43d55a2811d2dbb, you can get x4 generation speed in transformers with torch compile and static cache!

Is this expected to work with llava-next?

ArthurZucker commented:
Feel free to open an issue if it doesnt work

aliencaocao commented:
I have tried and it dont work because the vision tower changes the shape of inputs after encoding to patches. Also, it doesnt work for bnb 4 bits

ArthurZucker commented:
bnb is a different issue, torch.compile might not support this (int8 yes). 
For the encoder part cc @NielsRogge could be nice

aliencaocao commented:
we are using NF4 for bnb

nxphi47 commented:
> It is mostly this: https://gist.github.com/ArthurZucker/af34221def212259b43d55a2811d2dbb, you can get x4 generation speed in transformers with torch compile and static cache!

Just checking have you added this to `model.generate`, or we still have to follow your script there to use static KV cache?

ArthurZucker commented:
@gante is working on this here #29374
